cc/core/Mat.cc

```diff
@@ -167,6 +186,16 @@ NAN_METHOD(Mat::New) {
 		self->setNativeProps(mat);
 	}
 	self->Wrap(info.Holder());
+    
+// only valid for 3.1.0+
+#if CV_VERSION_MINOR > 0
+    // I *think* New should be called in JS thread where cv::mat has been created async,
+    // so a good place to rationalise memory
```

Comment:
Yep the JS constructor is always called from the main thread. Async workers don't initialize JS objects.

Reply:
:).  I found that MANY of the actual allocations of cv::mat were in other threads, so we have to rationalise here once the cv:mat has been passed back to us.  I was surprised about how many allocations where involved in a very simple process...  it's worth putting a printf in the custom function just to understand what's going on in terms of cv:mat allocations and de-allocations.